### PR TITLE
Reporting API: apply `hideColumns` recursively to nested values by setting `hideColumnsRecursively=1`

### PR DESCRIPTION
**Pull request by mattab**

### CODE REVIEW

File: core/DataTable/Filter/ColumnDelete.php [hunk 1]
```diff
@@ -181,8 +181,7 @@ protected function removeColumnsFromTable(&$table)
                 unset($table[$index][$column]);
             }
 
-            // Restore me in Piwik 4
-            //$this->removeColumnsFromTable($row);
+            $this->removeColumnsFromTable($row);
```
Comment (sgiehl) [inline core/DataTable/Filter/ColumnDelete.php]: That actually doesn't work for arrays, as `$row` is no reference, but a 'copy', so the filter isn't applied correctly for nested arrays. I'll apply a fix

**Comment by tsteur** [comment]
That be actually a very big (breaking) change. Are we sure we'd want to do this and are we clear on all that it means for all the different API calls? I'm not sure I'd want this merged actually and rather close it I would say.

**Comment by sgiehl** [comment]
@tsteur what about having an additional option for those filters to filter recursively? could maybe also be exposed as api parameter or something like that?

**Comment by tsteur** [comment]
@sgiehl that could work. Although generally not seeing that very often and then we need to add test cases and support it etc. Generally that would work though 👍

**Comment by sgiehl** [comment]
maybe @mattab can say something about the intention of that PR? Not sure what it would be needed for 🤔

**Comment by mattab** [comment]
The goal of the PR was to be able to filter columns in the output of the Live APIs, where currently showColumns/hideColumns aren't applied, and all data is always returned for all visitors. it was requested a few times over the years to be able to remove/show only some columns in these Live API outputs, as they are wordy/large by default.

**Comment by tsteur** [comment]
If it's easy to do we can introduce one additional general API parameter to enable this behaviour but we would not want this by default.

**Comment by mattab** [comment]
Another idea: could we otherwise only offer this specific behavior (by default) to Live APIs only? because it's only useful for Live APIs afaik so not really needed to add this feature to our API as a whole maybe...

**Comment by sgiehl** [comment]
@mattab @tsteur I've updated the PR. Now there is an additional parameter `hideColumnsRecursively`, which can be set to `1` to enable the removal of nested columns. For Live API methods that parameter defaults to `1`. All others will assume `0`.

Note: for `showColumns` that might not work as expected.

**Comment by tsteur** [comment]
@sgiehl can you add one system test for the Live API?

**Comment by sgiehl** [comment]
added a test

**Pull request merged**